### PR TITLE
feat: upgrade transparent proxy

### PR DESCRIPTION
**Pull request by sakulstra**
The code change here is that the `_admin` on TransparentProxy is immutable so things like `changeAdmin` are no longer possible.

The rational is that instead of changing the admin, one should change the owner of the admin.
This way one can omit a storage read which adds up quite a bit if one gas proxies talking to proxies.

---

This change in behavior also introduces some breaking changes though:
1. the admin of a TransparentProxy can no longer change
2. the admin is part of the constructor and thus the create2 address prediction changes
3. the ProxyAdmin no longer initializes with owner as msg.sender, but now also accepts a initialOwner parameter, so create2 prediction changes here as well.

OZ enforces a new ProxyAdmin per Proxy, which for us is a bit suboptimal.
Therefore instead of using their exact code, I had to copy it and change the constructor to accept a ProxyAdmin as input.

### CODE REVIEW

File: src/contracts/transparent-proxy/TransparentProxyFactoryBase.sol [hunk 1]
```diff
@@ -16,21 +16,16 @@ import {ProxyAdmin} from './ProxyAdmin.sol';
  **/
 abstract contract TransparentProxyFactoryBase is ITransparentProxyFactory {
   /// @inheritdoc ITransparentProxyFactory
-  function create(
-    address logic,
-    address admin,
-    bytes calldata data
-  ) external returns (address) {
+  function create(address logic, ProxyAdmin admin, bytes calldata data) external returns (address) {
```
Comment (kyzia551) [inline src/contracts/transparent-proxy/TransparentProxyFactoryBase.sol:19]: why do we use `ProxyAdmin` here?

File: src/contracts/transparent-proxy/TransparentProxyFactoryBase.sol [hunk 2]
```diff
@@ -16,21 +16,16 @@ import {ProxyAdmin} from './ProxyAdmin.sol';
...
-    emit ProxyCreated(proxy, logic, admin);
+    emit ProxyCreated(proxy, logic, address(admin));
     return proxy;
   }
 
   /// @inheritdoc ITransparentProxyFactory
   function createProxyAdmin(address adminOwner) external returns (address) {
-    address proxyAdmin = address(new ProxyAdmin());
-    IOwnable(proxyAdmin).transferOwnership(adminOwner);
+    address proxyAdmin = address(new ProxyAdmin(adminOwner));
```
Comment (kyzia551) [inline src/contracts/transparent-proxy/TransparentProxyFactoryBase.sol:28]: and here return it as address, not `ProxyAdmin`

File: src/contracts/transparent-proxy/TransparentProxyFactoryBase.sol [hunk 3]
```diff
@@ -39,23 +34,22 @@ abstract contract TransparentProxyFactoryBase is ITransparentProxyFactory {
...
-    external
-    returns (address)
-  {
-    address proxyAdmin = address(new ProxyAdmin{salt: salt}());
-    IOwnable(proxyAdmin).transferOwnership(adminOwner);
+  function createDeterministicProxyAdmin(
+    address adminOwner,
+    bytes32 salt
+  ) external returns (address) {
+    address proxyAdmin = address(new ProxyAdmin{salt: salt}(adminOwner));
```
Comment (kyzia551) [inline src/contracts/transparent-proxy/TransparentProxyFactoryBase.sol:52]: and here

File: src/contracts/transparent-proxy/ProxyAdmin.sol
Comment (kyzia551) [inline src/contracts/transparent-proxy/ProxyAdmin.sol:1]: Maybe we should just import from oz directly? As we have 0 diff atm

File: src/contracts/transparent-proxy/TransparentProxyFactoryBase.sol
Comment (kyzia551) [inline src/contracts/transparent-proxy/TransparentProxyFactoryBase.sol:1]: Following the pattern, similiar to `ProxyAdmin`, maybe we can ditch IOwnable, it's not used anywhere outside I think, and use full `Ownable`

File: src/contracts/transparent-proxy/TransparentUpgradeableProxy.sol [hunk 4]
```diff
@@ -1,137 +1,126 @@
...
- * implementation. If the admin tries to call a function on the implementation it will fail with an error that says
- * "admin cannot fallback to proxy target".
+ * that call matches the {ITransparentUpgradeableProxy-upgradeToAndCall} function exposed by the proxy itself.
+ * 2. If the admin calls the proxy, it can call the `upgradeToAndCall` function but any other call won't be forwarded to
+ * the implementation. If the admin tries to call a function on the implementation it will fail with an error indicating
+ * the proxy admin cannot fallback to the target implementation.
+ *
+ * These properties mean that the admin account can only be used for upgrading the proxy, so it's best if it's a
+ * dedicated account that is not used for anything else. This will avoid headaches due to sudden errors when trying to
+ * call a function from the proxy implementation. For this reason, the proxy deploys an instance of {ProxyAdmin} and
```
Comment (kyzia551) [inline src/contracts/transparent-proxy/TransparentUpgradeableProxy.sol:43]: does not deploys

File: src/contracts/transparent-proxy/TransparentProxyFactoryBase.sol (hunk 1 above)
Comment (sakulstra) [reply src/contracts/transparent-proxy/TransparentProxyFactoryBase.sol:19]: it's a bit debatable, that's why i asked for careful review 😅 

My thinking was that if passing a non ProxyAdmin here would potentiall break things. It's simply assumed to pass a proxy admin. While an interface ofc is no perfect enforcement, the intention is to at least highlight what should be passed.

File: src/contracts/transparent-proxy/ProxyAdmin.sol
Comment (sakulstra) [reply src/contracts/transparent-proxy/ProxyAdmin.sol:1]: This:
`import {ITransparentUpgradeableProxy} from './TransparentUpgradeableProxy.sol';` is the diff.

I don't see a way without copying.

File: src/contracts/transparent-proxy/TransparentProxyFactoryBase.sol
Comment (sakulstra) [reply src/contracts/transparent-proxy/TransparentProxyFactoryBase.sol:1]: kk removed. As we have oz as a dependency, we should probably also remove most things from oz-common, but will defer to another pr.

**Comment by sakulstra** [comment]
@kyzia551 just to explain the approach here (as perhaps it was not as self explanatory as i though it was 😅).

The "original" version of the contract creates a ProxyAdmin.
As I changed the constructor to **expect** a proxy admin, as a precaution i changed the type to "ProxyAdmin", so you have to - with full consciousness  - cast to ProxyAdmin the input.

The rest I did not change, to cause as little changes as possible.

**Pull request merged**